### PR TITLE
common: release g_ceph_context before returns

### PR DESCRIPTION
**Pull request by tchaikov**
http://tracker.ceph.com/issues/17762

### CODE REVIEW

**Comment by badone** [comment]
LGTM

**Comment by dachary** [comment]
testing manually with the reproducer to verify it fixes the issue

**Comment by tchaikov** [comment]
- http://pulpito.ceph.com/kchai-2016-11-04_10:40:58-rados-wip-kefu-testing---basic-mira/
- http://pulpito.ceph.com/kchai-2016-11-05_15:35:54-rados-wip-kefu-testing---basic-mira/

failed tests are caused by #10049

**Comment by tchaikov** [comment]
`Log::entry()` is flushing when the program terminates, that's why the producer crashes.

**Comment by liewegas** [comment]
This looks right to me...

**Comment by tchaikov** [comment]
@liewegas i am changing this PR so the `global_init()` will return a `unique_ptr` which calls `g_ceph_context->put()` in its dtor. and also fixes some other `g_ceph_context` leakage.

File: src/global/global_init.cc [hunk 1]
```diff
@@ -325,6 +326,10 @@ void global_init(std::vector < const char * > *alt_def_args,
     output_ceph_version();
 
   g_ceph_context->crush_location.init_on_startup();
+
+  using Deleter = std::function<void(CephContext*)>;
+  return std::unique_ptr<CephContext, Deleter>{g_ceph_context,
+                                               [](CephContext *p) {p->put();}};
```
Comment (cbodley) [inline src/global/global_init.cc]: since `CephContext` is a ref-counted type, i think `boost::intrusive_ptr` is a more natural fit than `std::unique_ptr` - see src/rgw/rgw_main.cc for an example

**Comment by dachary** [comment]
jenkins test this please (eio now ignored in master)

**Comment by dachary** [comment]
jenkins test this please (asok now fixed in master)

**Comment by dachary** [comment]
<pre>
  9%] Performing configure step for 'rocksdb_ext'
CMake Error: The source directory "/home/jenkins-build/build/workspace/ceph-pull-requests/src/rocksdb" does not appear to contain CMakeLists.txt.
Specify --help for usage, or press the help button on the CMake GUI.
make[3]: *** [src/rocksdb_ext-prefix/src/rocksdb_ext-stamp/rocksdb_ext-configure] Error 1
make[2]: *** [src/CMakeFiles/rocksdb_ext.dir/all] Error 2
make[2]: *** Waiting for unfinished jobs....
[  9%] Building CXX object src/kv/CMakeFiles/kv_objs.dir/MemDB.cc.o
[  9%] Building CXX object src/kv/CMakeFiles/kv_objs.dir/RocksDBStore.cc.o
/home/jenkins-build/build/workspace/ceph-pull-requests/src/kv/RocksDBStore.cc:13:24: fatal error: rocksdb/db.h: No such file or directory
</pre>

**Comment by tchaikov** [comment]
retest this please.

File: src/global/global_init.cc (hunk 1 above)
Comment (tchaikov) [inline src/global/global_init.cc]: @cbodley comment addressed and repushed.

**Comment by tchaikov** [comment]
@athanatos mind taking a look?

**Comment by athanatos** [comment]
I like it!

**Comment by liewegas** [comment]
LGTM!  much better

**Comment by dachary** [comment]
@tchaikov excellent :-) In the commit message for f4c4f42 s/quites/quits/ + s/and ./and /. If at all possible it would be good to split  f4c4f42 into several so that it's more readable (the unit.h change for instance).

**Comment by tchaikov** [comment]
@dachary will do.

**Comment by tchaikov** [comment]
changelog

- just fixed the typos in commit message of f4c4f42

**Comment by tchaikov** [comment]
retest this please

**Comment by tchaikov** [comment]
./bin/unittest_erasure_code_plugin_jerasure' segfault. rerunning.

File: src/test/erasure-code/TestErasureCodePluginJerasure.cc [hunk 2]
```diff
@@ -61,24 +60,6 @@ TEST(ErasureCodePlugin, factory)
   }
 }
 
-int main(int argc, char **argv)
```
Comment (dachary) [inline src/test/erasure-code/TestErasureCodePluginJerasure.cc]: this is different from test/unit.h and needs to be adapted

**Comment by tchaikov** [comment]
changelog

- do not take a refcount when constructing the `intrusive_ptr<>`.

@liewegas , sorry! we need to reschedule a rados qa test.

File: src/test/erasure-code/TestErasureCodePluginJerasure.cc (hunk 2 above)
Comment (dachary) [inline src/test/erasure-code/TestErasureCodePluginJerasure.cc]: my bad, it was an autotools leftover

**Comment by tchaikov** [comment]
crushtool times out.

File: src/tools/crushtool.cc [hunk 3]
```diff
@@ -274,6 +274,9 @@ int main(int argc, const char **argv)
   auto cct = global_init(NULL, env_args, CEPH_ENTITY_TYPE_CLIENT,
 			 CODE_ENVIRONMENT_UTILITY,
 			 CINIT_FLAG_NO_DEFAULT_CONFIG_FILE);
+  // crushtool times out occasionally when quits. so do not
+  // release the g_ceph_context.
+  cct->get();
   common_init_finish(g_ceph_context);
```
Comment (dachary) [inline src/tools/crushtool.cc]: It would be nice to comment on the root cause of this timeout. A leak there is harmless but knowing the root cause would be good to check if that's because of a problem that may be shared by other parts of the code, maybe with side effects that are more subtle and more difficult to diagnose afterwards. Am I making sense ?

Comment (tchaikov) [inline src/tools/crushtool.cc]: @dachary i tried to reproduce this problem in a heavy loaded machine, the crushtool times out twice out of 200 runs. it could take more than 6 seconds or 16 seconds to complete the check. and i try to catch this by wrap the crushtool with a python script which launches gdb to print out the backtrace of all threads if it times out but the gdb reports that the debugged process is a zombie and cannot be attached.

in short, 1) the root cause is still unknown. 2) the leak is not a regression. i just make it more obvious.

Comment (dachary) [inline src/tools/crushtool.cc]: Ok

**Comment by tchaikov** [comment]
changelog

- `unit.h`: pass 0 instead of `CINIT_FLAG_NO_DEFAULT_CONFIG_FILE`. as we are also using the `ceph.conf` to customize the behavior of unit tests.

File: src/test/unit.h [hunk 4]
```diff
@@ -32,10 +34,19 @@
  * initialization for you.
  */
 int main(int argc, char **argv) {
-  std::vector<const char*> args;
-  global_init(NULL, args, CEPH_ENTITY_TYPE_CLIENT, CODE_ENVIRONMENT_UTILITY,
-	      CINIT_FLAG_NO_DEFAULT_CONFIG_FILE);
```
Comment (dachary) [inline src/test/unit.h]: CINIT_FLAG_NO_DEFAULT_CONFIG_FILE may be here so that tests are not influenced by an existing ceph installation on the developer machine (i.e. no attempt to read /etc/ceph/ceph.conf even if it's available). This is just a thought: I did not actually verify that's going to happen but ... that's what I thought.

**Comment by tchaikov** [comment]
tested at 

- http://pulpito.ceph.com/kchai-2016-11-24_15:35:58-rados-master---basic-mira/ without the "unit.h: pass 0 instead of CINIT_FLAG_NO_DEFAULT_CONFIG_FILE" change
- http://pulpito.ceph.com/kchai-2016-11-24_15:35:58-rados-master---basic-mira/ with the "unit.h: pass 0 instead of CINIT_FLAG_NO_DEFAULT_CONFIG_FILE" change

File: src/test/unit.h (hunk 4 above)
Comment (dachary) [inline src/test/unit.h]: right

**Comment by dillaman** [comment]
FYI -- this change caused rbd-nbd to immediately seg fault. I am fixing under http://tracker.ceph.com/issues/18070
